### PR TITLE
[feature] introduce device_options.filtered_cache to allow attributes to be filtered out from the cache

### PR DESCRIPTION
**Pull request by sjorge**
Sometimes It's not desirable for an attribute to end up in the cache, be it personal preference or something that makes little sense.

This PR allows to specify a per-device list of attributes to not be cached, but they will be published as usual. We already have `filtered_attributes` and `filtered_optimistic`, so a new `filtered_cache` was introduced to keep naming consistent.

We simply pass this list of attributes to State.set() as a new parameter and merge this with the already present hard-coded list *dontCacheProperties*,

### CODE REVIEW

**Comment by Koenkk** [comment]
Out of interest, can you share your use case for this?

**Comment by sjorge** [comment]
OK now that I am at my computer, let me write a more concrete example.

- You have a ubisys switch or bulb, you consume those publishes in node-red and translate those to homekit.
   - having access to state, brightness, color, ... on every publish is very handy here. So setting cache_state=true and cache_state_persistent=true.
 - You have a script to configure the switches/bulbs
   - subscribe to the device's topic
   - execute a /get for power_on_behavior, color_options ... 
   - wait for publish
     - compare to desired state and execute a /set to correct it to the desired state

After the script is done if the power_on_behavior, color_options ... are set in filtered_cache they will:
- not be present in the on disk cache (as no need to store them)
- not present in other publishes that follow after

File: lib/state.ts [hunk 1]
```diff
@@ -71,13 +71,13 @@ class State {
...
-    set(entity: Group | Device, update: KeyValue, reason: string=null): KeyValue {
+    set(entity: Group | Device, update: KeyValue, reason: string=null, extraDontCacheProperties: string[]=[]): KeyValue {
         const fromState = this.state[entity.ID] || {};
         const toState = objectAssignDeep({}, fromState, update);
         const result = {...toState};
 
         for (const property of Object.keys(toState)) {
-            if (dontCacheProperties.find((p) => property.match(p))) {
+            if (dontCacheProperties.concat(extraDontCacheProperties).find((p) => property.match(p))) {
                 delete toState[property];
```
[the remark this answers is not in the document]
Comment (Koenkk) [reply lib/state.ts:82]: They should be included since those are part of the state change, we just don't cache them.

File: lib/state.ts [hunk 2]
```diff
@@ -71,13 +71,14 @@ class State {
         return this.state[entity.ID] || {};
     }
 
-    set(entity: Group | Device, update: KeyValue, reason: string=null): KeyValue {
+    set(entity: Group | Device, update: KeyValue, reason: string=null,
+        extraDontCacheProperties: string[]=[]): KeyValue {
```
Comment (Koenkk) [inline lib/state.ts:75]: I believe it is not necessary to have an extra parameter for this. You can grab the options directly from the `entity`.

File: lib/state.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (sjorge) [reply lib/state.ts:82]: So the `to: toState` should be update to be `to: result` right ?

Or we should operate on `result` and use `this.state[entity.ID] = result;`, I'm not 100% sure whats the right one here.

File: lib/util/settings.schema.json [hunk 3]
```diff
@@ -812,23 +812,32 @@
...
           "type": "array",
           "items": {
             "type": "string"
           },
-          "examples": ["temperature", "battery", "action"],
-          "title": "Filtered publish attributes",
-          "description": "Filter attributes from publish payload."
+          "examples": ["action", "input_actions"],
+          "title": "Filtered attributes from cache",
+          "description": "Filter attributes from being added to the cache."
```
Comment (Koenkk) [inline lib/util/settings.schema.json:831]: I think `Filter attributes from being added to the cache, this prevents the attribute from being in the published payload when the value didn't change` is better

Comment (sjorge) [reply lib/util/settings.schema.json:831]: This is indeed a much clearer description!

File: lib/state.ts (hunk 2 above)
Comment (sjorge) [reply lib/state.ts:75]: That's even cleaner! And of course entity is passed in so there is no reason not too! Will update and retest on my end before push.

File: lib/state.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (sjorge) [reply lib/state.ts:82]: Done, gonna retest before push

**Comment by sjorge** [comment]
Gonna retest everything and push the changes in a bit.

**Comment by sjorge** [comment]
I somehow lost my tests when rebasing! Let me look into that.

File: lib/state.ts [hunk 4]
```diff
@@ -75,7 +75,7 @@ class State {
         const fromState = this.state[entity.ID] || {};
         const toState = objectAssignDeep({}, fromState, update);
         const newCache = {...toState};
-        const entityDontCacheProperties = entity.options.filtered_cache ? entity.options.filtered_cache : [];
+        const entityDontCacheProperties = entity.options.filtered_cache || [];
```
Comment (sjorge) [inline lib/state.ts:78]: Interesting, I had not considered this style of check.
Yet I use it all the time in node-red.

File: test/controller.test.js [hunk 5]
```diff
@@ -492,7 +492,7 @@ describe('Controller', () => {
 
     it('Publish entity state attribute_json output filtered cache', async () => {
         await controller.start();
-        settings.set(['experimental', 'output'], 'attribute_and_json');
+        settings.set(['advanced', 'output'], 'attribute_and_json');
```
Comment (sjorge) [inline test/controller.test.js:495]: Probably also want to get the other one above where I copied and modified this test from then?

Comment (Koenkk) [reply test/controller.test.js:495]: This is fine, in this way we keep test coverage of the legacy option.

**Comment by Koenkk** [comment]
Thanks as always 😄  🏅

**Comment by lolorc** [comment]
@sjorge, this implements a blacklist to prevent attributes from being cached, ever thought of a whitelist as well ? I think it could be convenient. and what about regular expressions ? ;-)

**Comment by sjorge** [comment]
I've never needed a passlist for things like this, however I don't think there is a technical reason why we can't have one.
cc. @Koenkk do you see any reason, aside from we probably need to rename filtered_* to pass_* and block_* I guess to make it more consistent.

**Comment by Koenkk** [comment]
I think this is fine, maybe we can support the regexes then you are able to do this (I don't want to make too many options for such specific features)

**Comment by sjorge** [comment]
So loop over the filter_cache entries for each property of the payload, re-match it and if it matches delete it?
How would we differentiate between regex and normal absolute match though.

**Comment by lolorc** [comment]
"string".constructor == String vs /re/.constructor == RegExp ?

@sjorge, you're concatening const dontCacheProperties which is an array full of regexp with the entity filtered_attributes setting.
The initial code already uses the match method, meaning there're no absolute match at the moment with dev, using filtered_attributes = [ "e" ] should be deleting every attribute with a "e" from the cache.

Might be easier to update the documentation and mention the list is made of regexp: to remove "temp" from cache it has to be written "^temp$".

**Comment by sjorge** [comment]
Looks like you're right, though just like the other two filtered_ is this was also using forEach but it is indeed using find.

I think the easiest might be to just switch the other 2 to find() instead of forEach() and update the docs.

**Pull request merged**